### PR TITLE
Implement grades filter

### LINKED ISSUE #5: Level filter on the map
## User story

As a user, I can filter problems by grade on the map.

## Demo

https://www.loom.com/share/1753930ded49434aa47bb4f543786d51

Note: If user selects for instance "7a" to "7b", you should actually filter from "7a" to "7b+" (notice the **+**)
I made this decision to avoid asking the user to select the grade in a huge list with all the grades "7a", "7a+", "7b", "7b+" etc

## iOS code 

- [mapbox filter](https://github.com/boolder-org/boolder-ios/blob/97cd4cb19483fc3461c8275e1299f638d6329567/Boolder/UI/Map/MapboxViewController.swift#L667-L672)
- [grade picker UI](https://github.com/boolder-org/boolder-ios/blob/97cd4cb19483fc3461c8275e1299f638d6329567/Boolder/UI/Map/FiltersView.swift#L21-L59)
- [grade range](https://github.com/boolder-org/boolder-ios/blob/97cd4cb19483fc3461c8275e1299f638d6329567/Boolder/Models/GradeRange.swift#L15-L23)

## Comments

Since there isn't any filters bar yet (like on iOS), let's find a way to make a simple button.
You can have a look at how I did it on the [web version](https://www.boolder.com/fr/map).

### PR DESCRIPTION
**Pull request by wang-li**
This commit adds a filter that allows to only display the boulders of a given grade range.

https://github.com/boolder-org/boolder-android/assets/8343416/08dcd05d-26f8-48ad-91e8-4f935c051dee

Fixes #5

### CODE REVIEW

File: app/build.gradle.kts [hunk 1]
```diff
@@ -33,12 +29,21 @@ android {
...
+
+    buildFeatures {
+        viewBinding = true
+        compose = true
+    }
+
     compileOptions {
-        sourceCompatibility = JavaVersion.VERSION_11
-        targetCompatibility = JavaVersion.VERSION_11
+        sourceCompatibility = JavaVersion.VERSION_17
```
Comment (nmondollot) [inline app/build.gradle.kts:39]: How many users are using a version between 11 and 16?
(meaning: how many curent and future users will we lose?)

File: app/src/main/java/com/boolder/boolder/domain/model/GradeRange.kt [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        GradeRange.BEGINNER -> "${getString(R.string.grade)} 1 -> 3"
+        GradeRange.LEVEL1 -> "${getString(R.string.grade)} 1"
+        GradeRange.LEVEL2 -> "${getString(R.string.grade)} 2"
+        GradeRange.LEVEL3 -> "${getString(R.string.grade)} 3"
+        GradeRange.LEVEL4 -> "${getString(R.string.grade)} 4"
+        GradeRange.LEVEL5 -> "${getString(R.string.grade)} 5"
+        GradeRange.LEVEL6 -> "${getString(R.string.grade)} 6"
+        GradeRange.LEVEL7 -> "${getString(R.string.grade)} 7"
+        GradeRange.LEVEL8 -> "${getString(R.string.grade)} 8"
+        else -> "${gradeRange.min} -> ${gradeRange.max}"
```
Comment (nmondollot) [inline app/src/main/java/com/boolder/boolder/domain/model/GradeRange.kt:40]: Let's use the unicode character : →
It's nicer

**Review by nmondollot — COMMENTED** [review]
Looking great! 🤩

**Comment by nmondollot** [comment]
I found a small bug: the app crashes when I select only one grade, for instance grade "5a" to grade "5a"
log: https://gist.github.com/nmondollot/8113fa3987e306f2010f9249cca0513f

File: app/build.gradle.kts [hunk 3]
```diff
@@ -8,23 +8,19 @@ plugins {
...
 android {
     namespace = "com.boolder.boolder"
-    compileSdk = 32
+    compileSdk = 33
 
     defaultConfig {
         applicationId = "com.boolder.boolder"
         minSdk = 21
-        targetSdk = 32
+        targetSdk = 33
```
Comment (nmondollot) [inline app/build.gradle.kts:16]: what does this do?

File: app/build.gradle.kts [hunk 4]
```diff
@@ -8,23 +8,19 @@ plugins {
 
 android {
     namespace = "com.boolder.boolder"
-    compileSdk = 32
+    compileSdk = 33
```
Comment (nmondollot) [inline app/build.gradle.kts:11]: what does this do?

File: app/build.gradle.kts (hunk 1 above)
Comment (nmondollot) [inline app/build.gradle.kts:39]: what does this do?
does it have an impact on users?

File: app/build.gradle.kts (hunk 3 above)
Comment (wang-li) [reply app/build.gradle.kts:16]: This is the targeted android SDK version, or the intended version android version that the app should run on. It is usually better to keep it in sync with the `compileSdk` version.

File: app/build.gradle.kts (hunk 4 above)
Comment (wang-li) [reply app/build.gradle.kts:11]: This is the version of the android SDK that is used to compile the application. The update was needed in order to use the last version of Jetpack compose.

Otherwise, some errors like the following are happening when compiling:
```
Dependency 'androidx.compose.material3:material3:1.0.1' requires libraries and applications that
depend on it to compile against version 33 or later of the
Android APIs.

:app is currently compiled against android-32.

Recommended action: Update this project to use a newer compileSdk
of at least 33, for example 33.

Note that updating a library or application's compileSdk (which
allows newer APIs to be used) can be done separately from updating
targetSdk (which opts the app in to new runtime behavior) and
minSdk (which determines which devices the app can be installed
on).
```
☝️ Some precisions are included about the `targetSdk` and the `compileSdk` 😉

File: app/build.gradle.kts (hunk 1 above)
Comment (wang-li) [reply app/build.gradle.kts:39]: These lines defines the available Java features available to the project, especially about the byte code. This was needed by some of the added libraries. This should not have impact on the users, as the compilation step should take care of desugaring the code for the older devices.

File: app/src/main/java/com/boolder/boolder/domain/model/GradeRange.kt (hunk 2 above)
Comment (wang-li) [reply app/src/main/java/com/boolder/boolder/domain/model/GradeRange.kt]: I was a bit hesitant to use this character as it looks a bit weird with the current font 😅 I'll let you judge whether you find it better or not

**Comment by wang-li** [comment]
> I found a small bug: the app crashes when I select only one grade, for instance grade "5a" to grade "5a"
log: https://gist.github.com/nmondollot/8113fa3987e306f2010f9249cca0513f

Great catch! I forgot to test that edge case, and it was actually leading to an empty grade range 😬 My fault as I forgot that the `subList()` function had an exclusive end index... Now it is fixed in the second commit 👍

File: app/src/main/java/com/boolder/boolder/domain/model/GradeRange.kt (hunk 2 above)
Comment (nmondollot) [reply app/src/main/java/com/boolder/boolder/domain/model/GradeRange.kt]: Ugh, not great indeed 😬 

Maybe it's better to use "4a to 5a"?
("4a à 5a" in French)

**Comment by nmondollot** [comment]
Thanks for the fix and the explanations! 🙏

File: app/src/main/java/com/boolder/boolder/domain/model/GradeRange.kt (hunk 2 above)
Comment (nmondollot) [reply app/src/main/java/com/boolder/boolder/domain/model/GradeRange.kt]: As a quick fix I'll use the ▸ character

Capture d’écran 2023-07-12 à 11 03 54
